Plugin/PaymentInformationManagement.php

```diff
@@ -0,0 +1,168 @@
+
+    public function __construct(
+        \Magento\Checkout\Model\Session $checkoutSession,
+        \Adyen\Payment\Helper\Data $adyenHelper,
+        \Adyen\Payment\Helper\Requests $adyenRequestHelper,
+        \Magento\Framework\Model\Context $context,
+        \Adyen\Payment\Gateway\Http\TransferFactory $transferFactory,
+        \Adyen\Payment\Gateway\Http\Client\TransactionPayment $transactionPayment,
+        \Adyen\Payment\Gateway\Validator\CheckoutResponseValidator $checkoutResponseValidator,
+        \Adyen\Payment\Gateway\Validator\ThreeDS2ResponseValidator $threeDS2ResponseValidator
+    )
+    {
+        $this->checkoutSession = $checkoutSession;
+        $this->adyenHelper = $adyenHelper;
+        $this->adyenRequestHelper = $adyenRequestHelper;
+        $this->context = $context;
+        $this->transferFactory = $transferFactory;
+        $this->transactionPayment = $transactionPayment;
+        $this->checkoutResponseValidator = $checkoutResponseValidator;
+        $this->threeDS2ResponseValidator = $threeDS2ResponseValidator;
+    }
+
+    /**
+     * @param \Magento\Checkout\Model\PaymentInformationManagement $subject
+     * @param $response
+     */
+    public function afterSavePaymentInformation(
+        \Magento\Checkout\Model\PaymentInformationManagement $subject,
+        $response
+    )
+    {
+        // Get payment and cart information from session
+        $quote = $this->checkoutSession->getQuote();
+        $payment = $quote->getPayment();
+
+        // in case payments response is already there we don't need to perform another payments call
+        // we indicate it with the placeOrder additionalInformation
+        if ($payment->getAdditionalInformation('placeOrder')) {
+            $payment->unsAdditionalInformation('placeOrder');
+            $quote->save();
+            return true;
+        }
+
+        // Init request array
+        $request = [];
+
+        // Merchant account data builder
+        $paymentMethod = $payment->getMethod();
+        $storeId = $quote->getStoreId();
+        $request = $this->adyenRequestHelper->buildMerchantAccountData($request, $paymentMethod, $storeId);
+
+        // Customer data builder
+        $customerId = $quote->getCustomerId();
+        $billingAddress = $quote->getBillingAddress();
+        $request = $this->adyenRequestHelper->buildCustomerData($request, $customerId, $billingAddress, $payment);
+
+        // Customer Ip data builder
+        $shopperIp = $quote->getRemoteIp();
+        $request = $this->adyenRequestHelper->buildCustomerIpData($request, $shopperIp);
+
+        // AddressDataBuilder
+        $shippingAddress = $quote->getShippingAddress();
+        $request = $this->adyenRequestHelper->buildAddressData($request, $billingAddress, $shippingAddress);
+
+        // PaymentDataBuilder
+        $currencyCode = $quote->getQuoteCurrencyCode();
+        $amount = $quote->getGrandTotal();
+        $reference = $quote->getId();
```

Comment:
$quote->getId() returns an int, so instead of having 000000007 as merchantReference we have 7, not sure if there is a way to get the orderNumber at this point of the flow

Reply:
yeah true... the problem is that we don't have an order at this point of the flow
maybe there is a reserved order number or something for the quotes - let's check it because the notification won't work in the current case